// @ionic/v1-toolkit/dist/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const cli_framework_1 = require("@ionic/cli-framework");
const commands_1 = require("./commands");
class IonicV1Namespace extends cli_framework_1.Namespace {
    getMetadata() {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            return {
                name: 'ionic-v1',
                summary: '',
            };
        });
    }
    getCommands() {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            return new cli_framework_1.CommandMap([
                ['build', () => tslib_1.__awaiter(this, void 0, void 0, function* () { return new commands_1.BuildCommand(this); })],
                ['serve', () => tslib_1.__awaiter(this, void 0, void 0, function* () { return new commands_1.ServeCommand(this); })],
            ]);
        });
    }
}
exports.IonicV1Namespace = IonicV1Namespace;
const namespace = new IonicV1Namespace();
function run(argv, env) {
    return tslib_1.__awaiter(this, void 0, void 0, function* () {
        yield cli_framework_1.execute({ namespace, argv, env });
    });
}
exports.run = run;
